pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,237 @@
+// SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package ssa
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/schemamutation"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// cachingUnstructuredExtractor is a caching implementation of v1.UnstructuredExtractor
+// using OpenAPI V3 discovery information.
+// TODO(erhan): try to upstream this code in kubernetes
+type cachingUnstructuredExtractor struct {
+	// added as field to not break the interface for other funcs, instantiated at each reconcile
+	ctx   context.Context
+	cache *GVKParserCache
+	dc    discovery.DiscoveryInterface
+}
+
+// NewCachingUnstructuredExtractor returns a new cachingUnstructuredExtractor
+func NewCachingUnstructuredExtractor(ctx context.Context, dc discovery.DiscoveryInterface, cache *GVKParserCache) (applymetav1.UnstructuredExtractor, error) {
+	return &cachingUnstructuredExtractor{
+		dc:    dc,
+		cache: cache,
+		ctx:   ctx,
+	}, nil
+}
+
+// Extract extracts the applied configuration owned by fieldManager from an unstructured object.
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+func discoveryPaths(ctx context.Context, rc rest.Interface) (map[string]OpenAPIGroupVersion, error) {
```

Comment:
this seems to be misnamed. I don't see it returning a path. I see it returning schema loaders by paths.

Why is it indexed by paths in the first place? Why not directly by gv? It is only used with gvs.

Reply:
terminology and most parts of the implementation was borrowed from upstream [client-go](https://github.com/kubernetes/client-go/blob/fb6ca040d7b7e2293898e45dc0bfd9181f553bcc/openapi/client.go#L43)

I think naming of `openapi.GroupVersion` is a bit confusing too. The underlying implementor is a  [handler3.OpenAPIDiscoveryGroupVersion](https://github.com/kubernetes/kube-openapi/blob/9e1beecbcb384a484fb5cbd41024f316d51fdfac/pkg/handler3/handler.go#L57) that is a not-necessarily-GV server path wrapped along with a rest client.
Below is the response of``kubectl get --raw '/openapi/v3'`
The returned paths are not necessarily GroupVersion, it includes Groups, and other non-GV paths.
Probably this is why it is called `Paths()` at upstream.

Can try to parse the paths and filter the GV ones, then store `schema.GroupVersion` as keys. Though calculating the path keys from GVs seemed simpler.

```json
{
  "paths": {
    ".well-known/openid-configuration": {
      "serverRelativeURL": "/openapi/v3/.well-known/openid-configuration?hash=73BCD53F499D27D225D70A0C79D84C6D96EF410A3E0B49E471D7CDB055230C486083EF11FC1C29E6AEACAF25B2D2B7E49297FE432F15A419E5991D503203C7F7"
    },
    "api": {
      "serverRelativeURL": "/openapi/v3/api?hash=180F6A8D2E4479013F92230D4093DE33088F2E1DA0C220298FBA4ED8129D327D3C24DFE5BBEA4945457A8A117FC948E89F3630AA06EDA06A868875A3DBE567FF"
    },
    "api/v1": {
      "serverRelativeURL": "/openapi/v3/api/v1?hash=2A0B7D1DD0AF45569217219A34EF7E93C9AA2026A039B7C199C8677779832FB2D69D1CB6DA68A0EBE54F0EFDDB2DB2ABFC47CE8DF8213B52ED5BF31C9AC42BC4"
    },
    "apis": {
      "serverRelativeURL": "/openapi/v3/apis?hash=9331F4E2B1319F29C18C93095ED9BF8790E36D9D56F8B1956CF3487402BF394D50DACB1FE91BBBAD05E15AD7B45AA524B76AADB67F243F33E454456842D5B053"
    },
    "apis/admissionregistration.k8s.io": {
      "serverRelativeURL": "/openapi/v3/apis/admissionregistration.k8s.io?hash=C45DE744067E28F6002B35EB8C3B9DDB574FAC161809AB3C864C3D5BE6593C5D02D4649C8570A94731C5ABAE848AE53324EFEAE97691F85ED88BBC5B190AC04C"
    },
    "apis/admissionregistration.k8s.io/v1": {
      "serverRelativeURL": "/openapi/v3/apis/admissionregistration.k8s.io/v1?hash=9FDF107E3C61B36E079CE1A262B2EB39437937BDB37CDE8052EE41A5FCAB091BC40ACE0AFF31EF209BEA0E8E473669641A93C0D6D6A5F12E8797978950D25B4B"
    },
    "...." :{
      "...": "REMOVED FOR BREVITY"
    },
    "logs": {
      "serverRelativeURL": "/openapi/v3/logs?hash=AC0F20C36395B3CF2711AE80C53F4304EA2F3951102EBF89A6B5E9F6AEAB60E1FF879C4384177B7AA61A890512A4806AB967C143579087A17D8C1CC29006153E"
    },
    "openid/v1/jwks": {
      "serverRelativeURL": "/openapi/v3/openid/v1/jwks?hash=FEB706252E8F3767A0134706D62378FCC2C589BFD1E9F11DB562F0B783AD4B4065ED5938AEC3D49302B8B0168D86538A13A475F4FCB5B5DCB96D419A4A30F59C"
    },
    "version": {
      "serverRelativeURL": "/openapi/v3/version?hash=329C6D6D665A9D8059D57E79584EBD07DD8EAC487FE776AE43627595767FD6C43ECEF82DFC3A30884961A18803AAAF5AB2C7F9E898E7A9410E31ECD1D9AC9EEA"
    }
  }
}
```